examples/branin/restore_state.py

```diff
@@ -0,0 +1,64 @@
+import os
+import logging
+import shutil
+
+from smac.facade.smac_facade import SMAC
+from smac.scenario.scenario import Scenario
+from smac.stats.stats import Stats
+from smac.runhistory.runhistory import RunHistory
+from smac.utils.io.traj_logging import TrajLogger
+
+"""
+This file runs SMAC and then restores the run with an extended computation
+budget. This will also work for runs that have crashed and are continued.
+"""
+
+def main():
+    # Initialize scenario, using runcount_limit as budget.
+    orig_scen_dict = {
+        'algo' : 'python cmdline_wrapper.py',
+        'paramfile' : 'param_config_space.pcs',
+        'run_obj' : 'quality',
+        'runcount_limit' : 25,
+        'deterministic' : True,
+        'output_dir' : 'restore_me'}
+    original_scenario = Scenario(orig_scen_dict)
+    smac = SMAC(scenario=original_scenario)
+    smac.optimize()
+
+    # Now the output is in the folder 'restore_me'
+    # We create a new scenario with an extended budget
+    new_scenario = Scenario(orig_scen_dict,
+            cmd_args={'runcount_limit': 50,      # overwrite these args
+                      'output_dir' : 'restored'})
+
+    # We load the runhistory, ...
+    rh_path = os.path.join(original_scenario.output_dir, "runhistory.json")
+    runhistory = RunHistory(aggregate_func=None)
+    runhistory.load_json(rh_path, new_scenario.cs)
+    # ... stats, ...
+    stats_path = os.path.join(original_scenario.output_dir, "stats.json")
+    stats = Stats(new_scenario)
+    stats.load(stats_path)
+    # ... and trajectory.
+    traj_path = os.path.join(original_scenario.output_dir, "traj_aclib2.json")
+    trajectory = TrajLogger.read_traj_aclib_format(
+        fn=traj_path, cs=new_scenario.cs)
+    incumbent = trajectory[-1]["incumbent"]
+    # Because we changed the output_dir, we might want to copy the old
+    # trajectory-file (runhistory and stats will be complete)
+    new_traj_path = os.path.join(new_scenario.output_dir, "traj_aclib2.json")
+    shutil.copy(traj_path, new_traj_path)
+
+    # Now we can initialize SMAC with the recovered objects and restore the
+    # state where we left off. By providing stats and a restore_incumbent, SMAC
+    # automatically detects the intention of restoring a state.
+    smac = SMAC(scenario=new_scenario,
```

Comment:
Does is it have to be so complicated? I expected (without actually trying it) that you can simply modify the already existing scenario object and then call the exisiting SMAC object again, e.g.

```
smac.solver.scenario.runcount_limit = 100 # overwrite limit
smac.optimize()
```

Reply:
Yes, that works perfectly fine. I added it to the file, but commented it out, because I understood the request in #321 as an example to recover the SMAC-output and start validation both in the python-code. I chose the "extended" version, so that there is some example-code on how to recover all the needed objects.